File: 2023/python/day4.py
#!/usr/bin/env python3

from aoc import *


def matches(ticket):
    number = iter(ticket.split())
    next(number)
    next(number)
    winners = set()
    while True:
        x = next(number)
        if x == "|":
            break
        winners.add(x)
    have = set(number)
    return len(winners & have)


def part1(input: str):
    points = 0
    for ticket in input.splitlines():
        if winners := matches(ticket):
            points += 1 << (winners - 1)
    return points


def part2(input: str):
    tickets = input.splitlines()
    copies = [1] * len(tickets)
    for i, ticket in enumerate(tickets):
        instances = copies[i]
        if winners := matches(ticket):
            for j in range(i + 1, i + winners + 1):
                copies[j] += instances

    return sum(copies)


if __name__ == "__main__":
    main()

File: 2023/src/day4.rs
use std::{collections::HashSet, iter::FromIterator};

use nom::{
    character::complete::{self, multispace1},
    multi::many1,
    sequence::preceded,
    IResult,
};

fn ints(input: &str) -> IResult<&str, Vec<u32>> {
    many1(preceded(multispace1, complete::u32))(input)
}

#[aoc(day4, part1)]
pub fn part1(input: &str) -> usize {
    input
        .lines()
        .map(|l| l.split(':').nth(1).unwrap())
        .map(|l| {
            let sets = l
                .split('|')
                .map(ints)
                .map(|x| x.unwrap().1)
                .map(HashSet::from_iter)
                .collect::<Vec<HashSet<u32>>>();
            let exp = sets[0].intersection(&sets[1]).count() as u32;
            2usize.pow(exp - 1)
        })
        .sum()
}

#[aoc(day4, part2)]
pub fn part2(input: &str) -> usize {
    let mut cards = input
        .lines()
        .map(|l| l.split(':').nth(1).unwrap())
        .map(|l| {
            let sets = l
                .split('|')
                .map(ints)
                .map(|x| x.unwrap().1)
                .map(HashSet::from_iter)
                .collect::<Vec<HashSet<u32>>>();
            sets[0].intersection(&sets[1]).count()
        })
        .map(|n| (n, 1))
        .collect::<Vec<(usize, usize)>>();

    let mut total = 0;
    for i in 0..cards.len() {
        let (won, copies) = cards[i];
        total += copies;
        for j in 1..=won {
            cards[i + j].1 += copies;
        }
    }
    total
}
